### PR TITLE
[GSA] Jaccard with reduceOnNeighbors

### PR DESCRIPTION
**Pull request by andralungu**
Gather: no-op
Sum: reduceOnNeighbors to create the set of neighbors
Apply: attach the computed values to the vertices
Scatter: getTriplets() + map - I kept the result in the vertex value for consistency with vertex-centric jaccard.

@vasia 
Once this gets the all green, I will do something similar for GSA Triangles.

### CODE REVIEW

File: src/main/java/example/GSAJaccardSimilarityMeasure.java [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+		// initialize the vertex values with empty sets
+		Graph<Long, Tuple2<HashSet<Long>, HashMap<Long, Double>>, NullValue> graph = Graph.fromDataSet(edges,
+				new MapFunction<Long, Tuple2<HashSet<Long>, HashMap<Long, Double>>>() {
+
+					@Override
+					public Tuple2<HashSet<Long>, HashMap<Long, Double>> map(Long id) throws Exception {
+						HashSet<Long> neighbors = new HashSet<Long>();
+						neighbors.add(id);
+
+						return new Tuple2<HashSet<Long>, HashMap<Long, Double>>(neighbors,
```
Comment (vasia) [inline src/main/java/example/GSAJaccardSimilarityMeasure.java]: Why Tuple2 of HashSets? What's the 2nd HashSet for?

File: src/main/java/example/GSAJaccardSimilarityMeasure.java [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+				}, env);
+
+		// Simulate GSA
+		// Gather: no-op in this case
+		// Sum: create the set of neighbors
+		DataSet<Tuple2<Long, Tuple2<HashSet<Long>, HashMap<Long, Double>>>> computedNeighbors =
+				graph.reduceOnNeighbors(new GatherNeighbors(), EdgeDirection.ALL);
+
+		// Apply: attach the computed values as vertex values
+		DataSet<Vertex<Long, Tuple2<HashSet<Long>, HashMap<Long, Double>>>> verticesWithNeighbors =
```
Comment (vasia) [inline src/main/java/example/GSAJaccardSimilarityMeasure.java]: I'm not sure what you're trying to do here.. After getting a <Id, HashSet> for each vertex in sum, you can simply joinWithVertices to attach the HashSets as vertex values.

File: src/main/java/example/GSAJaccardSimilarityMeasure.java (hunk 1 above)
Comment (andralungu) [inline src/main/java/example/GSAJaccardSimilarityMeasure.java]: At some point, you asked to have the entire set of computed Jaccard values in the vertex value. That way the skewed vertex will be overwhelmed, which is pretty much what we want to prove.

File: src/main/java/example/GSAJaccardSimilarityMeasure.java [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+					}
+				});
+
+		Graph<Long, Tuple2<HashSet<Long>, HashMap<Long, Double>>, NullValue> graphWithNeighbors =
+				Graph.fromDataSet(verticesWithNeighbors, edges, env);
+
+		// Scatter: compare neighbors; compute Jaccard
+		DataSet<Vertex<Long, HashMap<Long, Double>>> verticesWithJaccardValues =
+				graphWithNeighbors.getTriplets().flatMap(new ComputeJaccard())
+						.groupBy(0).reduce(new ReduceFunction<Vertex<Long, HashMap<Long, Double>>>() {
```
Comment (vasia) [inline src/main/java/example/GSAJaccardSimilarityMeasure.java]: This one is over-complicated as well. On each triplet you will have a pair of neighborhoods. With a single map, you can compare them and output a <srcId, trgId, jaccardValue> tuple, which will be the result.

File: src/main/java/example/GSAJaccardSimilarityMeasure.java (hunk 1 above)
Comment (vasia) [inline src/main/java/example/GSAJaccardSimilarityMeasure.java]: The entire neighborhood yes, not the jaccard values; these refer to the edges. Also, this one is the GSA implementation, not the vertex-centric --> we want to spread the computation over the edges (which is what the map on getTriplets does). Makes sense?

**Comment by andralungu** [comment]
PR Updated :)

**Pull request merged**